### PR TITLE
Add array-callback-return rule; Fix current code breaking the rule.

### PR DESCRIPTION
**Pull request by jorgefilipecosta**
## Description
Adds a rule to make sure Array functions that iterate on the array and should return something contain the return statement, otherwise, a forEach should probably be used instead.
A case like this was fixed at https://github.com/WordPress/gutenberg/pull/13953.

In PR https://github.com/WordPress/gutenberg/pull/13953 @aduth suggested the implementation of a lint rule to catch these cases. While trying to implement the rule and researching the best ways to do it, I noticed a rule like that already existed in the community and this PR is enabling it.

We are also changing the code to respect the new rule no observable changes should be expected.

## How has this been tested?
Observe the tests pass.
Do some smoke testing, adding blocks, uploading files, and verify everything still works as before.

### CODE REVIEW

File: packages/scripts/scripts/check-licenses.js [hunk 1]
```diff
@@ -258,6 +258,7 @@ modules.forEach( ( path ) => {
 					}, stringDetectedType );
 				}, detectedType );
 			}
+			return false;
```
Comment (aduth) [inline packages/scripts/scripts/check-licenses.js]: I think we should not alter whatever the current reduced value is, no?

```suggestion
			return detectedType;
```

The logic is hard to follow, but I wonder also if this shouldn't be implemented as an `Array#find`. That might be a bit more of a refactor than we want to do here, though.

File: packages/eslint-plugin/configs/es5.js [hunk 2]
```diff
@@ -1,6 +1,7 @@
 module.exports = {
 	rules: {
 		'array-bracket-spacing': [ 'error', 'always' ],
+		'array-callback-return': 'error',
```
Comment (aduth) [inline packages/eslint-plugin/configs/es5.js]: Should we note this rule change in CHANGELOG.md ? It may in-fact be considered a breaking change, which is not hugely problematic since the pending unreleased version of the module is already bumping the major.

**Review by aduth — CHANGES_REQUESTED** [review]

File: packages/eslint-plugin/configs/es5.js (hunk 2 above)
Comment (ntwb) [reply packages/eslint-plugin/configs/es5.js]: Yes, please do update the changelog with this change

Comment (jorgefilipecosta) [reply packages/eslint-plugin/configs/es5.js]: The changelog was added 👍

**Review by aduth — APPROVED** [review]
I agree this is a good addition 👍 

But I don't think it captures the original issue raised, where we use array methods which produce a value without assigning them to anything.

For example, I'd expect this to produce an error:

https://eslint.org/demo/#eyJ0ZXh0IjoibXlBcnJheS5tYXAoICggaSApID0+IGkgKTsiLCJvcHRpb25zIjp7InBhcnNlck9wdGlvbnMiOnsiZWNtYVZlcnNpb24iOjksInNvdXJjZVR5cGUiOiJzY3JpcHQiLCJlY21hRmVhdHVyZXMiOnt9fSwicnVsZXMiOnsiYXJyYXktY2FsbGJhY2stcmV0dXJuIjoyfSwiZW52Ijp7fX19

Since I'd expect the result of `Array#map` to be used in a variable assignment, return statement, or as an argument to some other function.

**Comment by aduth** [comment]
>But I don't think it captures the original issue raised, where we use array methods which produce a value without assigning them to anything.

I guess in some ways it indirectly addresses the issue, like as seen with the updated examples, where the implementation of an `Array#map` is such that we're erroneously treating it as an alias to `forEach` and thus not implement with the necessary `return` within the body.

Unless there's an easy way to capture the broader issue, I think it's a reasonable enough stop-gap solution to capture _most_ errors.

**Comment by jorgefilipecosta** [comment]
Thank you for the review @aduth. I agree this solution is not perfect but I expect that it captures most of the errors. If we find that it is enough later we may try a custom rule, a rule that verifies if the value in the map is relevant to the return seems something complex to do.
